## Preprocess

In [1]:
# Required Python Packages
import pandas as pd
import numpy as np
import plotly.plotly as py
import plotly.graph_objs as go
#py.sign_in('YOUR_PLOTLY_USER_NAME', 'API_KEY')
 
from sklearn.cross_validation import train_test_split
from sklearn.linear_model import LogisticRegression
from sklearn.ensemble import RandomForestClassifier
from sklearn import metrics

import matplotlib.pyplot as plt
import seaborn as sns

/home/daniel/anaconda/envs/siemens/lib/python3.6/site-packages/sklearn/cross_validation.py:41: DeprecationWarning:

This module was deprecated in version 0.18 in favor of the model_selection module into which all the refactored classes and functions are moved. Also note that the interface of the new CV iterators are different from that of this module. This module will be removed in 0.20.



In [2]:
dataset_dir = "dataset_merged.csv"
dataset = pd.read_csv(dataset_dir,sep=',')

In [3]:
dataset

BIB  COD  ERG    FAN GJAH      LUK  MYR        NUS  PKD  RAS     SIS  \
0     160  iii  www   80.0  iii      5.0  eee   800000.0  xxx    t   1.750   
1     153  uuu  aaa  200.0  rrr      0.0  mmm  2000000.0  xxx  NaN   0.290   
2       5  iii  www   96.0  iii     19.0   hh   960000.0   hh    t   0.000   
3       9  iii  www    0.0  iii    120.0  kkk        0.0  qqq  NaN   0.335   
4      40  iii  www  232.0  iii      0.0  mmm  2320000.0  xxx    f   0.500   
5       8  iii  aaa  160.0  rrr      0.0  kkk  1600000.0  jjj    f   0.500   
6     152  iii  www  276.0  iii      1.0  lll  2760000.0  lll  NaN   0.000   
7     176  iii  aaa  280.0  rrr    204.0  eee  2800000.0  jjj  NaN   0.250   
8     154  iii  www  220.0  iii    140.0  eee  2200000.0  xxx  NaN   0.290   
9      95  iii  www  320.0  iii     13.0  eee  3200000.0  jjj  NaN   1.085   
10    121  iii  www  280.0  iii      1.0  sss  2800000.0  jjj    f   1.500   
11    155  iii  aaa  160.0  rrr     20.0  lll  1600000.0  lll    f   0.000   
12     91  iii  aaa  340.0  rrr      0.0  sss  3400000.0  jjj    f   0.165   
13     91  iii  www  120.0  iii      1.0  yyy  1200000.0  xxx  NaN   1.000   
14     91  iii  www  120.0  iii      1.0  yyy  1200000.0  jjj    t   0.165   
15     19  iii  aaa  120.0  rrr      1.0   hh  1200000.0   hh  NaN   1.500   
16    125  iii  www  178.0  iii      0.0  kkk  1780000.0   ff    f   0.000   
17    128  iii  www  120.0  iii      0.0  eee  1200000.0  xxx    f   0.500   
18     37  iii  www  416.0  iii     21.0  eee  4160000.0  jjj    t   2.290   
19     31  iii  www  180.0  iii      0.0  kkk  1800000.0   ff    f   0.000   
20     25  iii  www  292.0  iii      0.0  fff  2920000.0  jjj    f   5.750   
21    164  uuu  www  288.0  iii      0.0  ooo  2880000.0  xxx  NaN   0.250   
22     78  iii  aaa    NaN  rrr     18.0  mmm        NaN  xxx  NaN   0.165   
23    131  iii  aaa   80.0  rrr      0.0  eee   800000.0  xxx  NaN   4.000   
24     27  iii  aaa  200.0  rrr    456.0  kkk  2000000.0   ff    f   4.500   
25    154  iii  www  120.0  iii     18.0  eee  1200000.0  xxx    f   1.000   
26     50  uuu  www  240.0  iii      0.0  mmm  2400000.0  jjj  NaN   1.000   
27     10  iii  www    0.0  iii     33.0  yyy        0.0  jjj  NaN   0.040   
28     40  iii  www  300.0  iii      0.0  fff  3000000.0  xxx  NaN   3.000   
29    155  iii  www  160.0  iii     40.0  mmm  1600000.0  xxx  NaN   0.040   
...   ...  ...  ...    ...  ...      ...  ...        ...  ...  ...     ...   
3670  164  iii  www  280.0  iii    824.0  sss  2800000.0  jjj  NaN   1.500   
3671   90  iii  www    0.0  iii      0.0  ggg        0.0   ff  NaN   0.000   
3672  152  iii  www  120.0  iii    300.0  kkk  1200000.0  xxx  NaN   1.500   
3673   40  iii  www    0.0  iii   7059.0   ee        0.0  xxx    t   0.625   
3674  164  iii  www  320.0  iii      0.0  kkk  3200000.0  jjj  NaN   5.000   
3675   60  iii  www   80.0  iii      0.0  mmm   800000.0  xxx  NaN   0.500   
3676  131  iii  www   21.0  iii  10561.0  sss   210000.0  xxx  NaN  11.000   
3677   13  iii  www  216.0  iii      0.0  kkk  2160000.0   ff  NaN   3.500   
3678  164  iii  www  491.0  iii      0.0  eee  4910000.0  jjj  NaN   0.875   
3679  126  iii  www  168.0  iii      0.0   ee  1680000.0  xxx    t   5.040   
3680  147  iii  www    0.0  iii    456.0  eee        0.0  xxx    t   2.000   
3681  130  iii  www  250.0  iii    730.0  fff  2500000.0  xxx    t   3.000   
3682  124  iii  www    0.0  iii   1000.0   hh        0.0   hh    t  18.000   
3683  101  iii  www   80.0  iii    278.0  yyy   800000.0  xxx  NaN   2.000   
3684   68  iii  www    0.0  iii    500.0   ee        0.0   ff  NaN   3.500   
3685  140  iii  www  515.0  iii    500.0  zzz  5150000.0  jjj  NaN   1.625   
3686    1  iii  www    0.0  iii      0.0  eee        0.0   ff  NaN  15.000   
3687   97  iii  www   80.0  iii     50.0   ee   800000.0  jjj    t   0.335   
3688  158  iii  www  360.0  iii      0.0  ooo  3600000.0  xxx  NaN   2.500   
3689  158  iii  www  290.0  

## Dealing with NaNs 
In order to pre-process the dataset, we need to check how many rows have NaNs and what attributes contain the most amount of NaNs 

In [4]:
dataset.isnull().sum()

BIB         0
COD         0
ERG        64
FAN       100
GJAH       64
LUK         0
MYR        66
NUS       100
PKD        66
RAS      2145
SIS         0
TOK         0
UIN        39
VOL         0
WET         0
KAT        39
XIN         0
Class       0
dtype: int64

As observed the RAS attribute has an overwhelming amount of NaNs and should be considered to deletion. But before, we compare how many rows are dropped if we only accept a certain amount of NaNs per row, or in other words, how many attributes are allowed to be missing in every row.

In [5]:
# We will compare the dataset shape in terms of rows dropped when dropping rows with N nans
dataset_clean = dataset.copy()
print("Dataset shape vs Number of NaNs allowed")
for nan in range(19):
    data = dataset_clean.dropna(thresh=(dataset_clean.shape[1]-nan))
    if dataset_clean.shape != None:
        print("Dataset shape {} with allowed number of NaNs {}.".format(data.shape,nan))

Dataset shape vs Number of NaNs allowed
Dataset shape (1463, 18) with allowed number of NaNs 0.
Dataset shape (3578, 18) with allowed number of NaNs 1.
Dataset shape (3616, 18) with allowed number of NaNs 2.
Dataset shape (3636, 18) with allowed number of NaNs 3.
Dataset shape (3636, 18) with allowed number of NaNs 4.
Dataset shape (3636, 18) with allowed number of NaNs 5.
Dataset shape (3652, 18) with allowed number of NaNs 6.
Dataset shape (3700, 18) with allowed number of NaNs 7.
Dataset shape (3700, 18) with allowed number of NaNs 8.
Dataset shape (3700, 18) with allowed number of NaNs 9.
Dataset shape (3700, 18) with allowed number of NaNs 10.
Dataset shape (3700, 18) with allowed number of NaNs 11.
Dataset shape (3700, 18) with allowed number of NaNs 12.
Dataset shape (3700, 18) with allowed number of NaNs 13.
Dataset shape (3700, 18) with allowed number of NaNs 14.
Dataset shape (3700, 18) with allowed number of NaNs 15.
Dataset shape (3700, 18) with allowed number of NaNs 16.
D

As seen above, dropping all NaNs is not an option, since we would have to remove about 50% of the dataset! So further considerations have to be made

## Deleting RAS

In [6]:
<img src="RAS_Target_Histogram.png">

SyntaxError: invalid syntax (<ipython-input-6-b273e38a175b>, line 1)

One of this considerations can be to delete the RAS attribute enterely. Since the attribute is missing ~50% of the times, it may indicate a malfunctionioning device or data pipeline. Furthermore, by observing the histogram, both values of RAS correlate positively, so we here assume that information lost by deleting RAS is minimal. Other approaches could be to build a classify to predict the missing values, but there are just too many values missing.

To anaylse this thesis, one can delete the RAS collumn enterely and compare the shape of the data set vs allowing a specific number of NaNs per row.

In [6]:
dataset_clean

BIB  COD  ERG    FAN GJAH      LUK  MYR        NUS  PKD  RAS     SIS  \
0     160  iii  www   80.0  iii      5.0  eee   800000.0  xxx    t   1.750   
1     153  uuu  aaa  200.0  rrr      0.0  mmm  2000000.0  xxx  NaN   0.290   
2       5  iii  www   96.0  iii     19.0   hh   960000.0   hh    t   0.000   
3       9  iii  www    0.0  iii    120.0  kkk        0.0  qqq  NaN   0.335   
4      40  iii  www  232.0  iii      0.0  mmm  2320000.0  xxx    f   0.500   
5       8  iii  aaa  160.0  rrr      0.0  kkk  1600000.0  jjj    f   0.500   
6     152  iii  www  276.0  iii      1.0  lll  2760000.0  lll  NaN   0.000   
7     176  iii  aaa  280.0  rrr    204.0  eee  2800000.0  jjj  NaN   0.250   
8     154  iii  www  220.0  iii    140.0  eee  2200000.0  xxx  NaN   0.290   
9      95  iii  www  320.0  iii     13.0  eee  3200000.0  jjj  NaN   1.085   
10    121  iii  www  280.0  iii      1.0  sss  2800000.0  jjj    f   1.500   
11    155  iii  aaa  160.0  rrr     20.0  lll  1600000.0  lll    f   0.000   
12     91  iii  aaa  340.0  rrr      0.0  sss  3400000.0  jjj    f   0.165   
13     91  iii  www  120.0  iii      1.0  yyy  1200000.0  xxx  NaN   1.000   
14     91  iii  www  120.0  iii      1.0  yyy  1200000.0  jjj    t   0.165   
15     19  iii  aaa  120.0  rrr      1.0   hh  1200000.0   hh  NaN   1.500   
16    125  iii  www  178.0  iii      0.0  kkk  1780000.0   ff    f   0.000   
17    128  iii  www  120.0  iii      0.0  eee  1200000.0  xxx    f   0.500   
18     37  iii  www  416.0  iii     21.0  eee  4160000.0  jjj    t   2.290   
19     31  iii  www  180.0  iii      0.0  kkk  1800000.0   ff    f   0.000   
20     25  iii  www  292.0  iii      0.0  fff  2920000.0  jjj    f   5.750   
21    164  uuu  www  288.0  iii      0.0  ooo  2880000.0  xxx  NaN   0.250   
22     78  iii  aaa    NaN  rrr     18.0  mmm        NaN  xxx  NaN   0.165   
23    131  iii  aaa   80.0  rrr      0.0  eee   800000.0  xxx  NaN   4.000   
24     27  iii  aaa  200.0  rrr    456.0  kkk  2000000.0   ff    f   4.500   
25    154  iii  www  120.0  iii     18.0  eee  1200000.0  xxx    f   1.000   
26     50  uuu  www  240.0  iii      0.0  mmm  2400000.0  jjj  NaN   1.000   
27     10  iii  www    0.0  iii     33.0  yyy        0.0  jjj  NaN   0.040   
28     40  iii  www  300.0  iii      0.0  fff  3000000.0  xxx  NaN   3.000   
29    155  iii  www  160.0  iii     40.0  mmm  1600000.0  xxx  NaN   0.040   
...   ...  ...  ...    ...  ...      ...  ...        ...  ...  ...     ...   
3670  164  iii  www  280.0  iii    824.0  sss  2800000.0  jjj  NaN   1.500   
3671   90  iii  www    0.0  iii      0.0  ggg        0.0   ff  NaN   0.000   
3672  152  iii  www  120.0  iii    300.0  kkk  1200000.0  xxx  NaN   1.500   
3673   40  iii  www    0.0  iii   7059.0   ee        0.0  xxx    t   0.625   
3674  164  iii  www  320.0  iii      0.0  kkk  3200000.0  jjj  NaN   5.000   
3675   60  iii  www   80.0  iii      0.0  mmm   800000.0  xxx  NaN   0.500   
3676  131  iii  www   21.0  iii  10561.0  sss   210000.0  xxx  NaN  11.000   
3677   13  iii  www  216.0  iii      0.0  kkk  2160000.0   ff  NaN   3.500   
3678  164  iii  www  491.0  iii      0.0  eee  4910000.0  jjj  NaN   0.875   
3679  126  iii  www  168.0  iii      0.0   ee  1680000.0  xxx    t   5.040   
3680  147  iii  www    0.0  iii    456.0  eee        0.0  xxx    t   2.000   
3681  130  iii  www  250.0  iii    730.0  fff  2500000.0  xxx    t   3.000   
3682  124  iii  www    0.0  iii   1000.0   hh        0.0   hh    t  18.000   
3683  101  iii  www   80.0  iii    278.0  yyy   800000.0  xxx  NaN   2.000   
3684   68  iii  www    0.0  iii    500.0   ee        0.0   ff  NaN   3.500   
3685  140  iii  www  515.0  iii    500.0  zzz  5150000.0  jjj  NaN   1.625   
3686    1  iii  www    0.0  iii      0.0  eee        0.0   ff  NaN  15.000   
3687   97  iii  www   80.0  iii     50.0   ee   800000.0  jjj    t   0.335   
3688  158  iii  www  360.0  iii      0.0  ooo  3600000.0  xxx  NaN   2.500   
3689  158  iii  www  290.0  

In [7]:
# We will compare the dataset shape in terms of rows dropped when dropping rows with N nans
dataset_clean = dataset.copy()
print("Dataset shape after removing RAS attribute vs Number of NaNs allowed")
del dataset_clean['RAS']

Dataset shape after removing RAS attribute vs Number of NaNs allowed


In [8]:
for nan in range(19):
    dataset_without_RAS = dataset_clean.dropna(thresh=(dataset_clean.shape[1]-nan))
    if dataset_clean.shape != None:
        print("Dataset shape {} with allowed number of NaNs {}.".format(dataset_without_RAS.shape,nan))

Dataset shape (3522, 17) with allowed number of NaNs 0.
Dataset shape (3598, 17) with allowed number of NaNs 1.
Dataset shape (3634, 17) with allowed number of NaNs 2.
Dataset shape (3636, 17) with allowed number of NaNs 3.
Dataset shape (3636, 17) with allowed number of NaNs 4.
Dataset shape (3636, 17) with allowed number of NaNs 5.
Dataset shape (3700, 17) with allowed number of NaNs 6.
Dataset shape (3700, 17) with allowed number of NaNs 7.
Dataset shape (3700, 17) with allowed number of NaNs 8.
Dataset shape (3700, 17) with allowed number of NaNs 9.
Dataset shape (3700, 17) with allowed number of NaNs 10.
Dataset shape (3700, 17) with allowed number of NaNs 11.
Dataset shape (3700, 17) with allowed number of NaNs 12.
Dataset shape (3700, 17) with allowed number of NaNs 13.
Dataset shape (3700, 17) with allowed number of NaNs 14.
Dataset shape (3700, 17) with allowed number of NaNs 15.
Dataset shape (3700, 17) with allowed number of NaNs 16.
Dataset shape (3700, 17) with allowed num

In [9]:
dataset_clean.isnull().sum()

BIB        0
COD        0
ERG       64
FAN      100
GJAH      64
LUK        0
MYR       66
NUS      100
PKD       66
SIS        0
TOK        0
UIN       39
VOL        0
WET        0
KAT       39
XIN        0
Class      0
dtype: int64

As seen, in the extreme of discarding all rows with Nans, we only lose (3700-3522) = 178 lines! So discarding RAS seems an appropriate solution.

## Dealing with more NaNs

One could just discard the rest of the NaNs and skip further considerations, but a closer look can be given to the lines containing NaNs. Bellow we will expose all the lines with Nans.

In [10]:
nans = lambda lines_with_nans: dataset_clean[dataset_clean.isnull().any(axis=1)]
nans(dataset_clean)

BIB  COD  ERG     FAN GJAH      LUK  MYR         NUS  PKD    SIS TOK  \
22     78  iii  aaa     NaN  rrr     18.0  mmm         NaN  xxx  0.165   f   
28     40  iii  www   300.0  iii      0.0  fff   3000000.0  xxx  3.000   t   
38    163  iii  aaa     0.0  rrr      0.0   ee         0.0  xxx  8.500   f   
86     17  iii  aaa   150.0  rrr      1.0  NaN   1500000.0  NaN  0.000   t   
108     1  rrr  NaN     NaN  NaN      0.0  NaN         NaN  NaN  0.000   f   
137    98  iii  www     0.0  iii      0.0  zzz         0.0  xxx  6.500   f   
139    40  uuu  www  1160.0  iii      0.0  kkk  11600000.0   ff  0.500   f   
157   158  iii  aaa   411.0  rrr      0.0  kkk   4110000.0  xxx  0.085   t   
159   119  iii  aaa     NaN  rrr      0.0   hh         NaN   hh  0.000   f   
164    12  iii  www   200.0  iii    105.0   hh   2000000.0   hh  0.000   t   
169   155  iii  aaa     0.0  rrr      1.0  kkk         0.0   ff  7.000   f   
196     1  rrr  NaN     NaN  NaN      0.0  NaN         NaN  NaN  0.000   f   
225    23  iii  aaa   200.0  rrr    141.0   ee   2000000.0  xxx  3.500   t   
228   161  iii  www   260.0  iii   1004.0   ee   2600000.0  xxx  0.040   f   
262    31  iii  www     NaN  iii    600.0  eee         NaN  xxx  2.250   f   
263   116  iii  www    73.0  iii    444.0  eee    730000.0   ff  4.750   f   
276   166  iii  www   300.0  iii      0.0  yyy   3000000.0  xxx  0.250   t   
301   147  iii  www   263.0  iii      0.0  sss   2630000.0  xxx  0.125   t   
318     1  rrr  NaN     NaN  NaN      0.0  NaN         NaN  NaN  0.000   f   
362   165  iii  www   380.0  iii   2010.0  mmm   3800000.0  xxx  0.250   f   
396   144  iii  www   240.0  iii    768.0  sss   2400000.0  xxx  1.000   f   
402     1  rrr  NaN     NaN  NaN      0.0  NaN         NaN  NaN  0.000   f   
419   104  iii  www     NaN  iii   5200.0   hh         NaN   hh  0.000   f   
421   152  iii  aaa   256.0  rrr     17.0  ggg   2560000.0  jjj  2.000   f   
432    77  iii  aaa     NaN  rrr      0.0  kkk         NaN   ff  0.165   t   
472   158  iii  www   290.0  iii   2279.0  zzz   2900000.0  xxx  5.000   t   
478    74  iii  www     NaN  iii  26726.0  sss         NaN  xxx  2.710   f   
485    60  iii  www     0.0  iii    340.0  NaN         0.0  NaN  0.540   f   
520     1  rrr  NaN     NaN  NaN      0.0  NaN         NaN  NaN  0.000   f   
549    98  iii  www     0.0  iii      0.0  zzz         0.0  xxx  6.500   f   
...   ...  ...  ...     ...  ...      ...  ...         ...  ...    ...  ..   
3088    1  rrr  NaN     NaN  NaN      0.0  NaN         NaN  NaN  0.000   f   
3117   98  iii  www     0.0  iii      0.0  zzz         0.0  xxx  6.500   f   
3137    1  rrr  NaN     NaN  NaN      0.0  NaN         NaN  NaN  0.000   f   
3165   31  iii  www     NaN  iii    600.0  eee         NaN  xxx  2.250   f   
3166  116  iii  www    73.0  iii    444.0  eee    730000.0   ff  4.750   f   
3183    1  rrr  NaN     NaN  NaN      0.0  NaN         NaN  NaN  0.000   f   
3223  144  iii  www   240.0  iii    768.0  sss   2400000.0  xxx  1.000   f   
3229    1  rrr  NaN     NaN  NaN      0.0  NaN         NaN  NaN  0.000   f   
3261  158  iii  www   290.0  iii   2279.0  zzz   2900000.0  xxx  5.000   t   
3267   74  iii  www     NaN  iii  26726.0  sss         NaN  xxx  2.710   f   
3302    1  rrr  NaN     NaN  NaN      0.0  NaN         NaN  NaN  0.000   f   
3331   98  iii  www     0.0  iii      0.0  zzz         0.0  xxx  6.500   f   
3351    1  rrr  NaN     NaN  NaN      0.0  NaN         NaN  NaN  0.000   f   
3379   31  iii  www     NaN  iii    600.0  eee         NaN  xxx  2.250   f   
3380  116  iii  www    73.0  iii    444.0  eee    730000.0   ff  4.750   f   
3397    1  rrr  NaN     NaN  NaN      0.0  NaN         NaN  NaN  0.000   f   
3437  144  iii  www   240.0  iii    768.0  sss   2400000.0  xxx  1.000   f   
3443    1  rrr  NaN     NaN  NaN      0.0  NaN         NaN  NaN  0.000   f   
3475  158  iii  www   290.0  iii   2279.0  zzz   2900000.0  xxx  5.000   t   
3481   74  iii  www     NaN 

By observaing the 179 lines with NaNs left, there is a clear pattern on the lines with most NaNs. Lines containing BIB = 1 and COD = rrr have an average of 6 NsNs! A thesis can be made wether these valyes have a correlation with a data pipeline malfunctioning or misslabeling and could require further research. But singe 5 NaNs per row is a rather large amount of information lost, we choose to remove these lines.

## Removing lines with BIB = 1 and COD = rrr due to mistake

In [11]:
dataset_clean = dataset_without_RAS[(dataset_without_RAS.ERG != 'rrr') & (dataset_without_RAS.BIB !=1)] 

In [12]:
dataset_clean

BIB  COD  ERG    FAN GJAH      LUK  MYR        NUS  PKD     SIS TOK  \
0     160  iii  www   80.0  iii      5.0  eee   800000.0  xxx   1.750   t   
1     153  uuu  aaa  200.0  rrr      0.0  mmm  2000000.0  xxx   0.290   f   
2       5  iii  www   96.0  iii     19.0   hh   960000.0   hh   0.000   f   
3       9  iii  www    0.0  iii    120.0  kkk        0.0  qqq   0.335   f   
4      40  iii  www  232.0  iii      0.0  mmm  2320000.0  xxx   0.500   t   
5       8  iii  aaa  160.0  rrr      0.0  kkk  1600000.0  jjj   0.500   t   
6     152  iii  www  276.0  iii      1.0  lll  2760000.0  lll   0.000   t   
7     176  iii  aaa  280.0  rrr    204.0  eee  2800000.0  jjj   0.250   f   
8     154  iii  www  220.0  iii    140.0  eee  2200000.0  xxx   0.290   f   
9      95  iii  www  320.0  iii     13.0  eee  3200000.0  jjj   1.085   f   
10    121  iii  www  280.0  iii      1.0  sss  2800000.0  jjj   1.500   f   
11    155  iii  aaa  160.0  rrr     20.0  lll  1600000.0  lll   0.000   f   
12     91  iii  aaa  340.0  rrr      0.0  sss  3400000.0  jjj   0.165   f   
13     91  iii  www  120.0  iii      1.0  yyy  1200000.0  xxx   1.000   f   
14     91  iii  www  120.0  iii      1.0  yyy  1200000.0  jjj   0.165   f   
15     19  iii  aaa  120.0  rrr      1.0   hh  1200000.0   hh   1.500   f   
16    125  iii  www  178.0  iii      0.0  kkk  1780000.0   ff   0.000   t   
17    128  iii  www  120.0  iii      0.0  eee  1200000.0  xxx   0.500   f   
18     37  iii  www  416.0  iii     21.0  eee  4160000.0  jjj   2.290   t   
19     31  iii  www  180.0  iii      0.0  kkk  1800000.0   ff   0.000   f   
20     25  iii  www  292.0  iii      0.0  fff  2920000.0  jjj   5.750   t   
21    164  uuu  www  288.0  iii      0.0  ooo  2880000.0  xxx   0.250   f   
22     78  iii  aaa    NaN  rrr     18.0  mmm        NaN  xxx   0.165   f   
23    131  iii  aaa   80.0  rrr      0.0  eee   800000.0  xxx   4.000   f   
24     27  iii  aaa  200.0  rrr    456.0  kkk  2000000.0   ff   4.500   f   
25    154  iii  www  120.0  iii     18.0  eee  1200000.0  xxx   1.000   f   
26     50  uuu  www  240.0  iii      0.0  mmm  2400000.0  jjj   1.000   t   
27     10  iii  www    0.0  iii     33.0  yyy        0.0  jjj   0.040   f   
28     40  iii  www  300.0  iii      0.0  fff  3000000.0  xxx   3.000   t   
29    155  iii  www  160.0  iii     40.0  mmm  1600000.0  xxx   0.040   t   
...   ...  ...  ...    ...  ...      ...  ...        ...  ...     ...  ..   
3669  107  iii  www   80.0  iii   1349.0   ee   800000.0  xxx   0.040   f   
3670  164  iii  www  280.0  iii    824.0  sss  2800000.0  jjj   1.500   f   
3671   90  iii  www    0.0  iii      0.0  ggg        0.0   ff   0.000   f   
3672  152  iii  www  120.0  iii    300.0  kkk  1200000.0  xxx   1.500   f   
3673   40  iii  www    0.0  iii   7059.0   ee        0.0  xxx   0.625   f   
3674  164  iii  www  320.0  iii      0.0  kkk  3200000.0  jjj   5.000   t   
3675   60  iii  www   80.0  iii      0.0  mmm   800000.0  xxx   0.500   f   
3676  131  iii  www   21.0  iii  10561.0  sss   210000.0  xxx  11.000   t   
3677   13  iii  www  216.0  iii      0.0  kkk  2160000.0   ff   3.500   f   
3678  164  iii  www  491.0  iii      0.0  eee  4910000.0  jjj   0.875   t   
3679  126  iii  www  168.0  iii      0.0   ee  1680000.0  xxx   5.040   t   
3680  147  iii  www    0.0  iii    456.0  eee        0.0  xxx   2.000   f   
3681  130  iii  www  250.0  iii    730.0  fff  2500000.0  xxx   3.000   f   
3682  124  iii  www    0.0  iii   1000.0   hh        0.0   hh  18.000   t   
3683  101  iii  www   80.0  iii    278.0  yyy   800000.0  xxx   2.000   f   
3684   68  iii  www    0.0  iii    500.0   ee        0.0   ff   3.500   f   
3685  140  iii  www  515.0  iii    500.0  zzz  5150000.0  jjj   1.625   t   
3687   97  iii  www   80.0  iii     50.0   ee   800000.0  jjj   0.335   t   
3688  158  iii  www  360.0  iii      0.0  ooo  3600000.0  xxx   2.500   t   
3689  158  iii  www  290.0  iii   2279.0  zzz  2900000.0  xxx   5.000   t   
36

In [13]:
dataset_clean.isnull().sum()

BIB       0
COD       0
ERG       0
FAN      36
GJAH      0
LUK       0
MYR       2
NUS      36
PKD       2
SIS       0
TOK       0
UIN      39
VOL       0
WET       0
KAT      39
XIN       0
Class     0
dtype: int64

Only 6 attributes now contain missing values, since MYR and PKD have 2 lines with missing values we may remove them without affecting the information of the dataset in a negative manner.

## Removing the 2 lines of NANS and finnally go fill

In [14]:
dataset_clean.dropna(subset = ['MYR', 'PKD'],inplace=True)

/home/daniel/anaconda/envs/siemens/lib/python3.6/site-packages/ipykernel/__main__.py:1: SettingWithCopyWarning:


A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy



In [25]:
dataset_clean

BIB  COD  ERG    FAN GJAH      LUK  MYR        NUS  PKD     SIS TOK  \
0     160  iii  www   80.0  iii      5.0  eee   800000.0  xxx   1.750   t   
1     153  uuu  aaa  200.0  rrr      0.0  mmm  2000000.0  xxx   0.290   f   
2       5  iii  www   96.0  iii     19.0   hh   960000.0   hh   0.000   f   
3       9  iii  www    0.0  iii    120.0  kkk        0.0  qqq   0.335   f   
4      40  iii  www  232.0  iii      0.0  mmm  2320000.0  xxx   0.500   t   
5       8  iii  aaa  160.0  rrr      0.0  kkk  1600000.0  jjj   0.500   t   
6     152  iii  www  276.0  iii      1.0  lll  2760000.0  lll   0.000   t   
7     176  iii  aaa  280.0  rrr    204.0  eee  2800000.0  jjj   0.250   f   
8     154  iii  www  220.0  iii    140.0  eee  2200000.0  xxx   0.290   f   
9      95  iii  www  320.0  iii     13.0  eee  3200000.0  jjj   1.085   f   
10    121  iii  www  280.0  iii      1.0  sss  2800000.0  jjj   1.500   f   
11    155  iii  aaa  160.0  rrr     20.0  lll  1600000.0  lll   0.000   f   
12     91  iii  aaa  340.0  rrr      0.0  sss  3400000.0  jjj   0.165   f   
13     91  iii  www  120.0  iii      1.0  yyy  1200000.0  xxx   1.000   f   
14     91  iii  www  120.0  iii      1.0  yyy  1200000.0  jjj   0.165   f   
15     19  iii  aaa  120.0  rrr      1.0   hh  1200000.0   hh   1.500   f   
16    125  iii  www  178.0  iii      0.0  kkk  1780000.0   ff   0.000   t   
17    128  iii  www  120.0  iii      0.0  eee  1200000.0  xxx   0.500   f   
18     37  iii  www  416.0  iii     21.0  eee  4160000.0  jjj   2.290   t   
19     31  iii  www  180.0  iii      0.0  kkk  1800000.0   ff   0.000   f   
20     25  iii  www  292.0  iii      0.0  fff  2920000.0  jjj   5.750   t   
21    164  uuu  www  288.0  iii      0.0  ooo  2880000.0  xxx   0.250   f   
22     78  iii  aaa    NaN  rrr     18.0  mmm        NaN  xxx   0.165   f   
23    131  iii  aaa   80.0  rrr      0.0  eee   800000.0  xxx   4.000   f   
24     27  iii  aaa  200.0  rrr    456.0  kkk  2000000.0   ff   4.500   f   
25    154  iii  www  120.0  iii     18.0  eee  1200000.0  xxx   1.000   f   
26     50  uuu  www  240.0  iii      0.0  mmm  2400000.0  jjj   1.000   t   
27     10  iii  www    0.0  iii     33.0  yyy        0.0  jjj   0.040   f   
28     40  iii  www  300.0  iii      0.0  fff  3000000.0  xxx   3.000   t   
29    155  iii  www  160.0  iii     40.0  mmm  1600000.0  xxx   0.040   t   
...   ...  ...  ...    ...  ...      ...  ...        ...  ...     ...  ..   
3669  107  iii  www   80.0  iii   1349.0   ee   800000.0  xxx   0.040   f   
3670  164  iii  www  280.0  iii    824.0  sss  2800000.0  jjj   1.500   f   
3671   90  iii  www    0.0  iii      0.0  ggg        0.0   ff   0.000   f   
3672  152  iii  www  120.0  iii    300.0  kkk  1200000.0  xxx   1.500   f   
3673   40  iii  www    0.0  iii   7059.0   ee        0.0  xxx   0.625   f   
3674  164  iii  www  320.0  iii      0.0  kkk  3200000.0  jjj   5.000   t   
3675   60  iii  www   80.0  iii      0.0  mmm   800000.0  xxx   0.500   f   
3676  131  iii  www   21.0  iii  10561.0  sss   210000.0  xxx  11.000   t   
3677   13  iii  www  216.0  iii      0.0  kkk  2160000.0   ff   3.500   f   
3678  164  iii  www  491.0  iii      0.0  eee  4910000.0  jjj   0.875   t   
3679  126  iii  www  168.0  iii      0.0   ee  1680000.0  xxx   5.040   t   
3680  147  iii  www    0.0  iii    456.0  eee        0.0  xxx   2.000   f   
3681  130  iii  www  250.0  iii    730.0  fff  2500000.0  xxx   3.000   f   
3682  124  iii  www    0.0  iii   1000.0   hh        0.0   hh  18.000   t   
3683  101  iii  www   80.0  iii    278.0  yyy   800000.0  xxx   2.000   f   
3684   68  iii  www    0.0  iii    500.0   ee        0.0   ff   3.500   f   
3685  140  iii  www  515.0  iii    500.0  zzz  5150000.0  jjj   1.625   t   
3687   97  iii  www   80.0  iii     50.0   ee   800000.0  jjj   0.335   t   
3688  158  iii  www  360.0  iii      0.0  ooo  3600000.0  xxx   2.500   t   
3689  158  iii  www  290.0  iii   2279.0  zzz  2900000.0  xxx   5.000   t   
36

## Filling the numerical attributes

In [15]:
dataset_clean.isnull().sum()

BIB       0
COD       0
ERG       0
FAN      36
GJAH      0
LUK       0
MYR       0
NUS      36
PKD       0
SIS       0
TOK       0
UIN      39
VOL       0
WET       0
KAT      38
XIN       0
Class     0
dtype: int64

Since FAN, NUS and UIN are numerical attributes with a few missing values, one can fill these gaps in many ways. Bu building a reverse classifier from the class target attribute, compute the mean and standard deviation of the attributes and fill them using a probabilistic like a guassian, but here we will only replace by the mean, and backtrack in a later point if the classification accuracy is poor.

In [16]:
dataset_clean.fillna(dataset_clean.mean(),inplace=True)
dataset_clean.shape

/home/daniel/anaconda/envs/siemens/lib/python3.6/site-packages/pandas/core/generic.py:4355: SettingWithCopyWarning:


A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy



(3564, 17)

In [17]:
dataset_clean

BIB  COD  ERG         FAN GJAH      LUK  MYR           NUS  PKD     SIS  \
0     160  iii  www   80.000000  iii      5.0  eee  8.000000e+05  xxx   1.750   
1     153  uuu  aaa  200.000000  rrr      0.0  mmm  2.000000e+06  xxx   0.290   
2       5  iii  www   96.000000  iii     19.0   hh  9.600000e+05   hh   0.000   
3       9  iii  www    0.000000  iii    120.0  kkk  0.000000e+00  qqq   0.335   
4      40  iii  www  232.000000  iii      0.0  mmm  2.320000e+06  xxx   0.500   
5       8  iii  aaa  160.000000  rrr      0.0  kkk  1.600000e+06  jjj   0.500   
6     152  iii  www  276.000000  iii      1.0  lll  2.760000e+06  lll   0.000   
7     176  iii  aaa  280.000000  rrr    204.0  eee  2.800000e+06  jjj   0.250   
8     154  iii  www  220.000000  iii    140.0  eee  2.200000e+06  xxx   0.290   
9      95  iii  www  320.000000  iii     13.0  eee  3.200000e+06  jjj   1.085   
10    121  iii  www  280.000000  iii      1.0  sss  2.800000e+06  jjj   1.500   
11    155  iii  aaa  160.000000  rrr     20.0  lll  1.600000e+06  lll   0.000   
12     91  iii  aaa  340.000000  rrr      0.0  sss  3.400000e+06  jjj   0.165   
13     91  iii  www  120.000000  iii      1.0  yyy  1.200000e+06  xxx   1.000   
14     91  iii  www  120.000000  iii      1.0  yyy  1.200000e+06  jjj   0.165   
15     19  iii  aaa  120.000000  rrr      1.0   hh  1.200000e+06   hh   1.500   
16    125  iii  www  178.000000  iii      0.0  kkk  1.780000e+06   ff   0.000   
17    128  iii  www  120.000000  iii      0.0  eee  1.200000e+06  xxx   0.500   
18     37  iii  www  416.000000  iii     21.0  eee  4.160000e+06  jjj   2.290   
19     31  iii  www  180.000000  iii      0.0  kkk  1.800000e+06   ff   0.000   
20     25  iii  www  292.000000  iii      0.0  fff  2.920000e+06  jjj   5.750   
21    164  uuu  www  288.000000  iii      0.0  ooo  2.880000e+06  xxx   0.250   
22     78  iii  aaa  163.627268  rrr     18.0  mmm  1.636273e+06  xxx   0.165   
23    131  iii  aaa   80.000000  rrr      0.0  eee  8.000000e+05  xxx   4.000   
24     27  iii  aaa  200.000000  rrr    456.0  kkk  2.000000e+06   ff   4.500   
25    154  iii  www  120.000000  iii     18.0  eee  1.200000e+06  xxx   1.000   
26     50  uuu  www  240.000000  iii      0.0  mmm  2.400000e+06  jjj   1.000   
27     10  iii  www    0.000000  iii     33.0  yyy  0.000000e+00  jjj   0.040   
28     40  iii  www  300.000000  iii      0.0  fff  3.000000e+06  xxx   3.000   
29    155  iii  www  160.000000  iii     40.0  mmm  1.600000e+06  xxx   0.040   
...   ...  ...  ...         ...  ...      ...  ...           ...  ...     ...   
3669  107  iii  www   80.000000  iii   1349.0   ee  8.000000e+05  xxx   0.040   
3670  164  iii  www  280.000000  iii    824.0  sss  2.800000e+06  jjj   1.500   
3671   90  iii  www    0.000000  iii      0.0  ggg  0.000000e+00   ff   0.000   
3672  152  iii  www  120.000000  iii    300.0  kkk  1.200000e+06  xxx   1.500   
3673   40  iii  www    0.000000  iii   7059.0   ee  0.000000e+00  xxx   0.625   
3674  164  iii  www  320.000000  iii      0.0  kkk  3.200000e+06  jjj   5.000   
3675   60  iii  www   80.000000  iii      0.0  mmm  8.000000e+05  xxx   0.500   
3676  131  iii  www   21.000000  iii  10561.0  sss  2.100000e+05  xxx  11.000   
3677   13  iii  www  216.000000  iii      0.0  kkk  2.160000e+06   ff   3.500   
3678  164  iii  www  491.000000  iii      0.0  eee  4.910000e+06  jjj   0.875   
3679  126  iii  www  168.000000  iii      0.0   ee  1.680000e+06  xxx   5.040   
3680  147  iii  www    0.000000  iii    456.0  eee  0.000000e+00  xxx   2.000   
3681  130  iii  www  250.000000  iii    730.0  fff  2.500000e+06  xxx   3.000   
3682  124  iii  www    0.000000  iii   1000.0   hh  0.000000e+00   hh  18.000   
3683  101  iii  www   80.000000  iii    278.0  yyy  8.000000e+05  xxx   2.000   
3684   68  iii  www    0.000000  iii    500.0   ee  0.000000e+00   ff   3.500   
3685  140  iii  www  515.000000  iii    500.0  zzz  5.150000e+06  jjj   1.625   
3687   97  iii  www   80.000000  iii 

In [18]:
dataset_clean.isnull().sum()

BIB       0
COD       0
ERG       0
FAN       0
GJAH      0
LUK       0
MYR       0
NUS       0
PKD       0
SIS       0
TOK       0
UIN       0
VOL       0
WET       0
KAT      38
XIN       0
Class     0
dtype: int64

# Dealing with KAT (last attribute)

There are many solutions one could deal with KAT, for now we will use the simplest one, replacing the missing values by the most frequent one, in this case 'ddd'.

In [19]:
dataset_clean.fillna("ddd",inplace=True)

/home/daniel/anaconda/envs/siemens/lib/python3.6/site-packages/pandas/core/frame.py:3035: SettingWithCopyWarning:


A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy



In [20]:
dataset_clean.isnull().sum()

BIB      0
COD      0
ERG      0
FAN      0
GJAH     0
LUK      0
MYR      0
NUS      0
PKD      0
SIS      0
TOK      0
UIN      0
VOL      0
WET      0
KAT      0
XIN      0
Class    0
dtype: int64

In [21]:
dataset_clean.shape

(3564, 17)

At this point we cleaned the data set from all the missinv values losing 136 rows and one collumn. In the next notebook we proceed to build a classifier.

In [22]:
dataset_clean.to_csv("dataset_clean.csv",index=False)

## PCA??/

In [9]:
import pandas as pd
import numpy as np
from sklearn.decomposition import PCA
pca = PCA(n_components=2)

In [8]:
!jupyter nbconvert Task\ 1\ Binary\ Classification\ -\ Preprocessing.ipynb --to html

[NbConvertApp] Converting notebook Task 1 Binary Classification - Preprocessing.ipynb to html
[NbConvertApp] Writing 394673 bytes to Task 1 Binary Classification - Preprocessing.html
